android/src/main/java/com/polidea/flutter_ble_lib/event/ConnectionStateStreamHandler.java

```diff
@@ -27,13 +27,22 @@ synchronized public void onCancel(Object o) {
 
     synchronized public void onNewConnectionState(final ConnectionStateChange connectionState) {
         if (eventSink != null) {
+            final ConnectionStateStreamHandler that = this;
             new Handler(Looper.getMainLooper()).post(new Runnable() {
                 @Override
                 public void run() {
-                    try {
-                        eventSink.success(connectionStateChangeJsonConverter.toJson(connectionState));
-                    } catch (JSONException e) {
-                        eventSink.error("-1", e.getMessage(), e.getStackTrace());
+                    final String json = connectionStateChangeJsonConverter.toJson(connectionState);
+
+                    synchronized (that) {
```

Comment:
I'm not sure if you actually want to synchronize across this block, but it seems like the easiest way to ensure we're free from race conditions.